contracts/EtomicSwapV2.sol

```diff
@@ -0,0 +1,483 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract EtomicSwapV2 {
+    enum MakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerSpent,
+        MakerRefunded
+    }
+
+    struct MakerPayment {
+        bytes20 paymentHash;
+        uint32 paymentLockTime;
+        MakerPaymentState state;
+    }
+
+    event MakerPaymentSent(bytes32 id);
+    event MakerPaymentSpent(bytes32 id, bytes32 secret);
+    event MakerPaymentRefundedTimelock(bytes32 id);
+    event MakerPaymentRefundedSecret(bytes32 id, bytes32 secret);
+
+    mapping(bytes32 => MakerPayment) public makerPayments;
+
+    enum TakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerApproved,
+        MakerSpent,
+        TakerRefunded
+    }
+
+    struct TakerPayment {
+        bytes20 paymentHash;
+        uint32 immediateRefundTime;
+        uint32 paymentLockTime;
+        TakerPaymentState state;
+    }
+
+    event TakerPaymentSent(bytes32 id);
+    event TakerPaymentApproved(bytes32 id);
+    event TakerPaymentSpent(bytes32 id, bytes32 secret);
+    event TakerPaymentRefundedSecret(bytes32 id, bytes32 secret);
+    event TakerPaymentRefundedTimelock(bytes32 id);
+
+    mapping(bytes32 => TakerPayment) public takerPayments;
+
+    address public immutable dexFeeAddress;
+
+    constructor(address feeAddress) {
+        require(feeAddress != address(0), "feeAddress must not be zero address");
+
+        dexFeeAddress = feeAddress;
+    }
+
+    function ethMakerPayment(
+        bytes32 id,
+        address taker,
+        bytes20 takerSecretHash,
+        bytes20 makerSecretHash,
+        uint32 paymentLockTime
+    ) external payable {
+        require(makerPayments[id].state == MakerPaymentState.Uninitialized, "Maker payment is already initialized");
+        require(taker != address(0), "Taker must not be zero address");
+        require(msg.value > 0, "ETH value must be greater than zero");
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                msg.value,
+                taker,
+                msg.sender,
+                takerSecretHash,
+                makerSecretHash,
+                address(0)
```

Comment:
do we really need to make zero address part of eth payment hash? I guess if removed we would need to duplicate a lot of functions, e.g. 2 spendMakerPayment functions, one for erc20 and one for eth.